### DISCUSSION TITLE
iOS example app

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#97] [Show and tell]
Implement a very basic iOS application using `whisper.cpp`

The [ggwave-objc](https://github.com/ggerganov/ggwave-objc) project can be used as a good starting point. It already provides the audio capture functionality. We just need to pass the captured data to `whisper.cpp`.

**Comment by ggerganov** [comment]
Just finished the initial implementation of the sample app and it is amazing!

https://user-images.githubusercontent.com/1991296/197378526-835929d3-8d87-4904-9d2f-f2c286fa38ad.mp4

Edit: and I had forgotten to enable the `Accelerate` framework.. after enabling it - it's x2.5 times faster compared to what you see in the video! 🚀

---

Edit: And here is the latest version using the `Accelerate` framework + `base.en` model:

https://user-images.githubusercontent.com/1991296/197385372-962a6dea-bca1-4d50-bf96-1d8c27b98c81.mp4

**Reply by cmp-nct** [reply]
Impressive

**Comment by yujinqiu** [comment]
great! will it  support stream mode ? @ggerganov

**Reply by ggerganov** [reply]
It's easy to add, but I will leave it as an exercise for someone with more experience in Objective-C and iOS apps

**Reply by ggerganov** [reply]
There is now basic streaming mode support added to the [whisper.objc](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc) application:

https://user-images.githubusercontent.com/1991296/204126266-ce4177c6-6eca-4bd9-bca8-0e46d9da2364.mp4

**Comment by siilats** [comment]
How do you enable the large project? it fails to allocate even if I manually enter the sizes

**Reply by ggerganov** [reply]
I haven't tried, but I believe you won't be able to load the `large` model because iOS apps have a memory limit of 2GB

**Comment by siilats** [comment]
It looks like there is an entitlement to get up to 16gb on iPad Pro

https://9to5mac.com/2021/06/25/apps-can-request-access-to-more-ram-with-ios-15-entitlement-exceeding-normal-system-memory-limits/amp/

It looks like the parameters in iOS app are hardcoded not read from the bin
file, right?

On Thu, Dec 8, 2022 at 12:19 AM Georgi Gerganov ***@***.***>
wrote:

> I haven't tried, but I believe you won't be able to load the large model
> because iOS apps have a memory limit of 2GB
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggerganov/whisper.cpp/discussions/97#discussioncomment-4340550>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI2ZY44CORSMBPSP3JQHF3WMGDW7ANCNFSM6AAAAAARQRYMOE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Keith Siilats ***@***.*** www.bytelogics.com Fax: 646-352-4705
Phone: +1 (646) 881-4378

**Reply by ggerganov** [reply]
The model size is not hardcoded. The only thing that is hardcoded in the current example is the name of the model file:

https://github.com/ggerganov/whisper.cpp/blob/832b4f34c9a155352c9f4c121faf92dcf759f86d/examples/whisper.objc/whisper.objc/ViewController.m#L53

You just have to change it to `ggml-large.bin` and make sure that the model is uploaded to the iPhone / iPad.

**Comment by siilats** [comment]
Yeah I’m getting Malloc error and the model type is not 5 even though I
changed it to large. It looks like it doesn’t get the 20 layers from the
file.

On Thu, Dec 8, 2022 at 10:59 AM Georgi Gerganov ***@***.***>
wrote:

> The model size is not hardcoded. The only thing that is hardcoded in the
> current example is the name of the model file:
>
>
> https://github.com/ggerganov/whisper.cpp/blob/832b4f34c9a155352c9f4c121faf92dcf759f86d/examples/whisper.objc/whisper.objc/ViewController.m#L53
>
> You just have to change it to ggml-large.bin and make sure that the model
> is uploaded to the iPhone.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggerganov/whisper.cpp/discussions/97#discussioncomment-4345716>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI2ZYZMHBYGRE34HGBND6DWMIOWXANCNFSM6AAAAAARQRYMOE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Keith Siilats ***@***.*** www.bytelogics.com Fax: 646-352-4705
Phone: +1 (646) 881-4378

**Reply by ggerganov** [reply]
I just tested it with iPhone 14 Simulator and it works.
I only changed the model filename to `ggml-large.bin` and added the model to the project:

<img width="1916" alt="Screenshot 2022-12-08 at 22 35 20" src="https://user-images.githubusercontent.com/1991296/206562587-1d27413b-04ba-4043-a042-bd370db1a272.png">

<img width="1916" alt="Screenshot 2022-12-08 at 22 36 55" src="https://user-images.githubusercontent.com/1991296/206562611-854c424d-c257-4471-8d84-3606e1c962b7.png">

**Comment by siilats** [comment]
Hi, I also ran it successfully on the Iphone 14 simulator but when I try on the actual phone I get this error, I added the large memory entitlement.
Details

The app “whisper.objc” on iPhone quit unexpectedly.
Domain: IDEDebugSessionErrorDomain
Code: 11
Failure Reason: Message from debugger: Terminated due to memory issue
User Info: {
    DVTErrorCreationDateKey = "2022-12-20 01:43:16 +0000";
    IDERunOperationFailingWorker = DBGLLDBLauncher;
}
--

Analytics Event: com.apple.dt.IDERunOperationWorkerFinished : {
    "device_model" = "iPhone14,3";
    "device_osBuild" = "16.1.1 (20B101)";
    "device_platform" = "com.apple.platform.iphoneos";
    "launchSession_schemeCommand" = Run;
    "launchSession_state" = 2;
    "launchSession_targetArch" = arm64;
    "operation_duration_ms" = 5083;
    "operation_errorCode" = 11;
    "operation_errorDomain" = IDEDebugSessionErrorDomain;
    "operation_errorWorker" = DBGLLDBLauncher;
    "operation_name" = IDEiPhoneRunOperationWorkerGroup;
    "param_consoleMode" = 0;
    "param_debugger_attachToExtensions" = 0;
    "param_debugger_attachToXPC" = 1;
    "param_debugger_type" = 5;
    "param_destination_isProxy" = 0;
    "param_destination_platform" = "com.apple.platform.iphoneos";
    "param_diag_MainThreadChecker_stopOnIssue" = 0;
    "param_diag_MallocStackLogging_enableDuringAttach" = 0;
    "param_diag_MallocStackLogging_enableForXPC" = 1;
    "param_diag_allowLocationSimulation" = 1;
    "param_diag_checker_tpc_enable" = 1;
    "param_diag_gpu_frameCapture_enable" = 0;
    "param_diag_gpu_shaderValidation_enable" = 0;
    "param_diag_gpu_validation_enable" = 0;
    "param_diag_memoryGraphOnResourceException" = 0;
    "param_diag_queueDebugging_enable" = 1;
    "param_diag_runtimeProfile_generate" = 0;
    "param_diag_sanitizer_asan_enable" = 0;
    "param_diag_sanitizer_tsan_enable" = 0;
    "param_diag_sanitizer_tsan_stopOnIssue" = 0;
    "param_diag_sanitizer_ubsan_stopOnIssue" = 0;
    "param_diag_showNonLocalizedStrings" = 0;
    "param_diag_viewDebugging_enabled" = 1;
    "param_diag_viewDebugging_insertDylibOnLaunch" = 1;
    "param_install_style" = 0;
    "param_launcher_UID" = 2;
    "param_launcher_allowDeviceSensorReplayData" = 0;
    "param_launcher_kind" = 0;
    "param_launcher_style" = 0;
    "param_launcher_substyle" = 0;
    "param_runnable_appExtensionHostRunMode" = 0;
    "param_runnable_productType" = "com.apple.product-type.application";
    "param_runnable_type" = 2;
    "param_testing_launchedForTesting" = 0;
    "param_testing_suppressSimulatorApp" = 0;
    "param_testing_usingCLI" = 0;
    "sdk_canonicalName" = "iphoneos16.2";
    "sdk_osVersion" = "16.2";
    "sdk_variant" = iphoneos;
}
--


System Information

macOS Version 13.0.1 (Build 22A400)
Xcode 14.2 (21534) (Build 14C18)
Timestamp: 2022-12-19T20:43:16-05:00

**Comment by siilats** [comment]
When I did analyse I got this:
<img width="1452" alt="image" src="https://user-images.githubusercontent.com/1158371/208563964-ffd868bb-4f67-4cf1-9ec6-8c0e1d432d30.png">

**Comment by brandonmcgraw** [comment]
Is it possible to process a transcription from a saved WAV file (vs. getting from the microphone) in your iOS implementation of Whisper?

**Reply by ggerganov** [reply]
The [whisper.swiftui](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.swiftui) example already supports loading WAV files - give it a try.

**Comment by amnike** [comment]
Does real-time transcription have a limit? I'm only reaching 400 seconds max then it stops.

Model = base

**Reply by ggerganov** [reply]
The `whisper.objc` is actually limited to 30 seconds:

https://github.com/ggerganov/whisper.cpp/blob/4a3f0d3fe9ff7c3ba2896dfe27cf39224456032b/examples/whisper.objc/whisper.objc/ViewController.h#L14

It's just a basic example app. It can be improved to support longer transcriptions, but it will make it more complicated.

**Reply by amnike** [reply]
Thanks! 

Also as new results are generated, the previous ones are captured. As the number of captured results increases, the processing time also increases significantly. Is there a way to reduce the number of previous results captured to improve performance?

**Reply by Akeel14** [reply]
> Thanks!
> 
> Also as new results are generated, the previous ones are captured. As the number of captured results increases, the processing time also increases significantly. Is there a way to reduce the number of previous results captured to improve performance?

I'm following this cuz also looking for a solution but I think it's for accuracy.
